common/define-grammar.js

```diff
@@ -234,7 +234,14 @@ module.exports = function defineGrammar(dialect) {
         previous,
         seq('export', '=', $.identifier, $._semicolon),
         seq('export', 'as', 'namespace', $.identifier, $._semicolon),
-        seq('export', 'default', $.function_signature)
+        seq('export', 'default', $.function_signature),
+        seq(
+          'export', 'default',
+          choice(
+            seq(optional("abstract"), $.class),
+            $.interface_declaration
```

Comment:
I would think that all of these (`function_signature`, `abstract class` and `interface`) would be covered by the existing `declaration` rule. Could we just allow that instead, to make the grammar simpler? I'd prefer that, as long as the generated parsers don't end up larger.

Reply:
👋 Apologies if I make no sense at all, I have no experience with language grammars and am just eyeballing the whole codebase here. But I found what seems like a definition for `declaration` https://github.com/tree-sitter/tree-sitter-typescript/blob/c3c75049bd87822acaba12b3cea9ccdd48b84ba6/tsx/src/grammar.json#L396 and it seems to include things such as `enum_declaration` as possible choices.

I then went to copy [an example enum declaration from the TypeScript docs](https://www.typescriptlang.org/docs/handbook/enums.html#numeric-enums) and tried to place it after `export default`:

```typescript
export default enum Direction {
  Up,
  Down,
  Left,
  Right,
}
```

Now when I try to compile this with TypeScript 4.0, it prints

```
Failed to compile.

./src/singletons.ts
  Line 9:15:  Parsing error: Expression expected
```

So if I'm understanding everything correctly, you're asking to allow all declarations after `export default`, but not all are valid in the language. Sorry again if I'm grossly misunderstanding something here!

Reply:
@underyx your assessment makes sense, although I view this grammar as being tailored *for the editor* instead of being compliant to TypeScript's semantics. Put another way, for me it would be fine to not have 100% sensical syntax as far as semantics are concerned if the code ends up highlighting and parsing nicely even when encountering semantic errors. IMO not being able to export enums is not a parsing concern per-se, it's more like a TypeScript implementation detail.

Reply:
@resolritter It would be great to finish this, now that the fix on the javascript grammar is merged. I haven't followed the details here, but I might give it a shot if you're not available.

Reply:
@mjambon please, go ahead

Reply:
@resolritter great. https://github.com/tree-sitter/tree-sitter-typescript/pull/159 takes care of it.